File: week-5/protoseo/14501.java
import java.io.*;

public class Main {

    static int n;
    static int answer;
    static int[] t;
    static int[] p;
    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        t = new int[n];
        p = new int[n];
        for (int i = 0; i < n; i++) {
            final String[] split = br.readLine().split(" ");
            t[i] = Integer.parseInt(split[0]);
            if (i + t[i] > n) {
                p[i] = 0;
                continue;
            }
            p[i] = Integer.parseInt(split[1]);
        }

        find(0, 0);
        System.out.println(answer);
    }

    static void find(int now, int sum) {
        if (now >= n) {
            answer = Math.max(answer, sum);
            return;
        }
        find(now + t[now], sum + p[now]);
        find(now + 1, sum);
    }
}

File: week-05/devxb/14501.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.max

var N = 0
var ans = 0
val arr: Array<Pair<Int, Int>> = Array(16) { 0 to 0 }
val dp: Array<Int> = Array(16) { 0 }

fun main() {
    readLine()
    recur(1)
    println(ans)
}

fun readLine() {
    BufferedReader(InputStreamReader(System.`in`)).use { reader ->
        N = reader.readLine().toInt()
        for (i in 1..N) {
            val line = reader.readLine().split(" ").map { it.toInt() }
            arr[i] = line[0] to line[1]
        }
    }
}

fun recur(idx: Int) {
    if (idx > N) {
        return
    }
    for (i in idx - 1 downTo 1) {
        val arg = i + arr[i].first - 1
        if (arg >= idx) {
            continue
        }
        dp[idx] = max(dp[idx], dp[i])
    }
    dp[idx] += arr[idx].second
    if (idx + arr[idx].first - 1 > N) {
        dp[idx] -= arr[idx].second
    }
    ans = max(dp[idx], ans)
    recur(idx + 1)
}
